axelor-base/src/main/java/com/axelor/apps/base/service/PartnerServiceImpl.java

```diff
@@ -785,35 +795,131 @@ public boolean isRegistrationCodeValid(Partner partner) {
+        throw new AxelorException(
+            TraceBackRepository.CATEGORY_MISSING_FIELD,
+            I18n.get(BaseExceptionMessage.PARTNER_INVALID_REGISTRATION_CODE));
+      }
+      boolean isValidRegistrationCode =
+          Beans.get(klass).computeRegistrationCodeValidity(registrationCode);
+      if (!isValidRegistrationCode) {
+        throw new AxelorException(
+            TraceBackRepository.CATEGORY_MISSING_FIELD,
+            I18n.get(BaseExceptionMessage.PARTNER_INVALID_REGISTRATION_CODE));
+      }
+    } catch (AxelorException e) {
+      TraceBackService.trace(e, String.valueOf(ResponseMessageType.ERROR));
       return false;
     }
-    int i = registrationCode.length() - 1;
-    while (i > -1) {
-      int number = Character.getNumericValue(registrationCode.charAt(i));
-      if (number < 0) {
-        i--;
-        continue;
-      }
-      if (!isOddNumber) {
-        number *= 2;
+
+    return true;
+  }
+
+  @Override
+  public Class<? extends RegistrationNumberValidation> getRegistrationNumberValidationClass(
+      Partner partner) {
+    Address mainAddress = partner.getMainAddress();
+    if (mainAddress == null || mainAddress.getCountry() == null) {
+      return null;
+    }
+    Country businessCountry = partner.getMainAddress().getCountry();
+    RegistrationNumberTemplate registrationNumberTemplate =
+        businessCountry.getRegistrationNumberTemplate();
+
+    if (registrationNumberTemplate != null) {
+      Class<? extends RegistrationNumberValidation> klass = null;
+      try {
+        String origin = registrationNumberTemplate.getValidationMethodSelect();
+        if (!Strings.isNullOrEmpty(origin)) {
+          klass = (Class<? extends RegistrationNumberValidation>) Class.forName(origin);
+        }
+      } catch (ClassNotFoundException e) {
+        TraceBackService.trace(e, String.valueOf(ResponseMessageType.ERROR));
+      } finally {
+        if (klass == null) {
+          return RegistrationNumberValidationDefault.class;
+        }
+        return klass;
       }
-      if (number < 10) {
-        sum += number;
-      } else {
-        number -= 10;
-        sum += number + 1;
+    }
+    return null;
+  }
+
+  @Override
+  public String getRegistrationCodeTitleFromTemplate(Partner partner) {
+    if (partner.getMainAddress() != null && partner.getMainAddress().getCountry() != null) {
+      Country businessCountry = partner.getMainAddress().getCountry();
+      RegistrationNumberTemplate registrationNumberTemplate =
+          businessCountry.getRegistrationNumberTemplate();
+      if (registrationNumberTemplate != null
+          && !StringUtils.isBlank(registrationNumberTemplate.getTitleToDisplay())) {
+        return registrationNumberTemplate.getTitleToDisplay();
       }
-      i--;
-      isOddNumber = !isOddNumber;
     }
-    return sum % 10 == 0;
+    return null;
+  }
+
+  @Override
+  public Map<String, Map<String, Object>> getPartnerTypeSelectAttrs(Partner partner) {
```

Comment:
We have the usual issues with this architecture : it is introducing a lot of boilerplate code while not creating easily reusable methods. For this case it could be better to introduce a `PartnerVisibilityService` with a method returning a boolean per field, and call directly `getRegistrationCodeTitleFromTemplate(partner)` from the controller.